lib/ProductOpener/Numbers.pm

```diff
@@ -0,0 +1,148 @@
+# Address: 21 rue des Iles, 94100 Saint-Maur des Fossés, France
+#
+# Product Opener is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+=head1 NAME
+
+ProductOpener::Numbers - normalize numbers sent as strings in various formats
+with different sets of separators for digit grouping and to indicate decimals.
+
+=head1 DESCRIPTION
+
+=cut
+
+package ProductOpener::Numbers;
+
+use utf8;
+use Modern::Perl '2017';
+use Exporter    qw< import >;
+
+use Log::Any qw($log);
+
+BEGIN
+{
+	use vars       qw(@ISA @EXPORT_OK %EXPORT_TAGS);
+	@EXPORT_OK = qw(
+
+		&remove_insignificant_digits
+		&convert_string_to_number
+
+		);    # symbols to export on request
+	%EXPORT_TAGS = (all => [@EXPORT_OK]);
+}
+
+use vars @EXPORT_OK ;
+
+=head1 FUNCTIONS
+
+=head2 remove_insignificant_digits($)
+
+Some apps send us nutrient values that they have stored internally as
+floating point numbers.
+
+So we get values like:
+
+2.9000000953674
+1.6000000238419
+0.89999997615814
+0.359999990463256
+2.5999999046326
+
+On the other hand, when we get values like 2.0, 2.50 or 2.500,
+we want to keep the trailing 0s.
+
+The goal is to keep the precision if it makes sense. The tricky part
+is that we do not know in advance how many significant digits we can have,
+it varies from products to products, and even nutrients to nutrients.
+
+The desired output is thus:
+
+2.9000000953674 -> 2.9
+1.6000000238419 -> 1.6
+0.89999997615814 -> 0.9
+0.359999990463256 -> 0.36
+2.5999999046326 -> 2.6
+2 -> 2
+2.0 -> 2.0
+2.000 -> 2.000
+2.0001 -> 2
+0.0001 -> 0.0001
+
+=cut
+
+
+sub remove_insignificant_digits($) {
+
+	my $value = shift;
+	
+	# Make the value a string
+	$value .= '';
+	
+	# Very small values may have been converted to scientific notation
+	
+	if ($value =~ /\.(\d*?[1-9]\d*?)0{3}/) {
+		$value = $`. '.' . $1;
+	}
+	elsif ($value =~ /([1-9]0*)\.0{3}/) {
+		$value = $`. $1;
+	}
+	elsif ($value =~ /\.(\d*)([0-8]+)9999/) {
+		$value = $`. '.' . $1 . ($2 + 1);
+	}
+	elsif ($value =~ /\.9999/) {
+		$value = $` + 1;
+	}
+	return $value;
+}
+
+
+=head2 convert_string_to_number($)
+
+Try to convert a number represented as a string to the actual number,
+by guessing which characters (spaces, commas, dots) are used as
+digit grouping separators or used to indicate decimals.
+
+=cut
+
+sub convert_string_to_number($) {
+	
+	my $value = shift;
+	
+	$value =~ s/(\d) (\d)/$1$2/g;
```

Comment:
Do you know how often this gets called? This, and other, regex replaces could be replaced by [`tr`](https://stackoverflow.com/a/3847062/11963).

Reply:
Thanks @hangy , I merged the PR too soon. It's called whenever a nutrient value is updated. So it could be called a dozen of times for one product. In an import of 1000 products, it could be called in the order of 10 000 times. It's only when writing data, so I'm guessing it may not be worth spending too much time to improve its performance. We can replace the regexes with tr (especially if it helps readability).

Reply:
ooh, interesting. for some reason i saw the slashes, and assumed tr involved regex overhead too. i only ever use s/// with sed.